File: scripts/optimize_svgs.py
#!/usr/bin/env python3
import os
import sys
from pathlib import Path
try:
    from scour import scour
    SCOUR_AVAILABLE = True
except ImportError:
    SCOUR_AVAILABLE = False
    print("Warning: scour library not available. Please install with: pip install scour")

def optimize_svg_with_scour(file_path):
    """Optimize a single SVG file using scour library."""
    try:
        # Get original file size
        original_size = os.path.getsize(file_path)
        
        # Read the original SVG content
        with open(file_path, 'r') as f:
            svg_content = f.read()
        
        # Optimize the SVG using scour
        if SCOUR_AVAILABLE:
            # Configure scour options for maximum optimization
            options = scour.generateDefaultOptions()
            options.remove_descriptions = True
            options.remove_descriptive_elements = True
            options.remove_metadata = True
            options.remove_titles = True
            options.remove_descriptions = True
            options.remove_metadata = True
            options.remove_descriptive_elements = True
            options.enable_viewboxing = True
            options.strip_xml_prolog = True
            options.remove_comments = True
            options.remove_ids = True
            options.indent_type = None
            options.shorten_ids = True
            
            optimized_content = scour.scourString(svg_content, options)
            
            # Write optimized content back to file
            with open(file_path, 'w') as f:
                f.write(optimized_content)
        else:
            # If scour is not available, just return original values
            optimized_size = original_size
            return original_size, optimized_size, 0.0
        
        # Get optimized file size
        optimized_size = os.path.getsize(file_path)
        
        # Calculate savings
        saved_bytes = original_size - optimized_size
        saved_percentage = (saved_bytes / original_size) * 100 if original_size > 0 else 0
        
        return original_size, optimized_size, saved_percentage
    except Exception as e:
        print(f"Error processing {file_path}: {str(e)}")
        return None, None, None

def main():
    # Get the icons/src directory path
    icons_src_dir = Path(__file__).parent.parent / 'icons' / 'src'
    
    if not icons_src_dir.exists():
        print(f"Directory {icons_src_dir} does not exist")
        return
        
    # Find all SVG files
    svg_files = list(icons_src_dir.glob('*.svg'))
    
    if not svg_files:
        print(f"No SVG files found in {icons_src_dir}")
        return
        
    print(f"Found {len(svg_files)} SVG files to optimize...")
    
    # Process each SVG file
    results = []
    for svg_file in svg_files:
        print(f"Optimizing {svg_file.name}...")
        original_size, optimized_size, saved_percentage = optimize_svg_with_scour(str(svg_file))
        if original_size is not None:
            results.append({
                'file': svg_file.name,
                'original_size': original_size,
                'optimized_size': optimized_size,
                'saved_percentage': saved_percentage
            })
            print(f"{svg_file.name}: {original_size} → {optimized_size} bytes ({saved_percentage:.2f}% saved)")
    
    # Print summary
    if results:
        total_original = sum(r['original_size'] for r in results)
        total_optimized = sum(r['optimized_size'] for r in results)
        total_saved = total_original - total_optimized
        total_percentage = (total_saved / total_original) * 100 if total_original > 0 else 0
        
        print('\n--- Summary ---')
        print(f"Total files processed: {len(results)}")
        print(f"Original size: {total_original} bytes")
        print(f"Optimized size: {total_optimized} bytes")
        print(f"Total saved: {total_saved} bytes ({total_percentage:.2f}%)")


if __name__ == "__main__":
    main()

File: scripts/optimize-svgs.js
const fs = require('fs').promises;
const path = require('path');
const { optimize } = require('svgo');

// Configuration for SVGO optimization
const svgoConfig = {
  multipass: true,
  plugins: [
    {
      name: 'preset-default',
      params: {
        overrides: {
          removeXMLProcInst: true,     // Remove XML declarations
          removeDoctype: true,          // Remove DOCTYPE
          removeComments: true,        // Remove comments
          removeMetadata: true,        // Remove metadata
          removeEditorsNSData: true,   // Remove editor data
          cleanupAttrs: true,          // Clean up attributes
          removeEmptyAttrs: true,      // Remove empty attributes
          removeHiddenElems: true,     // Remove hidden elements
          removeEmptyText: true,      // Remove empty text
          removeEmptyContainers: true, // Remove empty container elements
          minifyStyles: true,          // Minify styles
          removeUselessStrokeAndFill: true, // Remove useless stroke and fill
          removeUnusedNS: true,        // Remove unused namespaces
          cleanupNumericValues: true,  // Cleanup numeric values
          convertColors: true,         // Convert colors
          removeUnknownsAndDefaults: true, // Remove unknown elements and defaults
          removeNonInheritableGroupAttrs: true, // Remove non-inheritable group attributes
          removeUselessDefs: true,     // Remove useless defs
          cleanupEnableBackground: true, // Cleanup enableBackground
          removeRasterImages: false,    // Keep raster images
          mergePaths: true,            // Merge paths
          convertShapeToPath: true,     // Convert shapes to paths
          sortAttrs: true,            // Sort attributes
          removeDimensions: false,       // Keep viewBox
        }
      }
    }
  ]
};

// Function to optimize a single SVG file
async function optimizeSvgFile(filePath) {
  try {
    // Read the original file
    const originalContent = await fs.readFile(filePath, 'utf8');
    const originalSize = Buffer.byteLength(originalContent, 'utf8');
    
    // Optimize the SVG
    const result = optimize(originalContent, svgoConfig);
    const optimizedContent = result.data;
    const optimizedSize = Buffer.byteLength(optimizedContent, 'utf8');
    
    // Calculate savings
    const savedBytes = originalSize - optimizedSize;
    const savedPercentage = ((savedBytes / originalSize) * 100).toFixed(2);
    
    // Write the optimized content back to the file
    await fs.writeFile(filePath, optimizedContent, 'utf8');
    
    console.log(`${path.basename(filePath)}: ${originalSize} → ${optimizedSize} bytes (${savedPercentage}% saved)`);
    
    return {
      fileName: path.basename(filePath),
      originalSize,
      optimizedSize,
      savedBytes,
      savedPercentage: parseFloat(savedPercentage)
    };
  } catch (error) {
    console.error(`Error optimizing ${filePath}:`, error.message);
    return null;
  }
}

// Main function to process all SVG files
async function main() {
  const iconsSrcDir = path.join(__dirname, '..', 'icons', 'src');
  
  try {
    // Read all files in the directory
    const files = await fs.readdir(iconsSrcDir);
    
    // Filter for SVG files
    const svgFiles = files.filter(file => path.extname(file) === '.svg');
    
    if (svgFiles.length === 0) {
      console.log('No SVG files found in icons/src/');
      return;
    }
    
    console.log(`Found ${svgFiles.length} SVG files to optimize...\n`);
    
    // Process each SVG file
    const results = [];
    for (const file of svgFiles) {
      const filePath = path.join(iconsSrcDir, file);
      const result = await optimizeSvgFile(filePath);
      if (result) {
        results.push(result);
      }
    }
    
    // Calculate and display totals
    const totalOriginalSize = results.reduce((sum, r) => sum + r.originalSize, 0);
    const totalOptimizedSize = results.reduce((sum, r) => sum + r.optimizedSize, 0);
    const totalSavedBytes = totalOriginalSize - totalOptimizedSize;
    const totalSavedPercentage = ((totalSavedBytes / totalOriginalSize) * 100).toFixed(2);
    
    console.log('\n--- Summary ---');
    console.log(`Total files processed: ${results.length}`);
    console.log(`Original size: ${totalOriginalSize} bytes`);
    console.log(`Optimized size: ${totalOptimizedSize} bytes`);
    console.log(`Total saved: ${totalSavedBytes} bytes (${totalSavedPercentage}%)`);
    
  } catch (error) {
    console.error('Error reading directory:', error.message);
  }
}

// Run the script
main();